python_modules/dagster/dagster/_core/definitions/metadata/__init__.py

```diff
@@ -1129,3 +1132,62 @@ def __new__(
     def value(self):
         """Alias of `entry_data`."""
         return self.entry_data
+
+
+T_NamespacedMetadataEntries = TypeVar(
+    "T_NamespacedMetadataEntries", bound="NamespacedMetadataEntries"
+)
+
+
+class NamespacedMetadataEntries(ABC, BaseModel, frozen=True):
+    """Extend this class to define a set of metadata fields in the same namespace.
+
+    Supports splatting to a dictionary that can be placed inside a metadata argument along with
+    other dictionary-structured metadata.
+
+    .. code-block:: python
+
+        my_metadata: NamespacedMetadataEntries = ...
+        return MaterializeResult(metadata={**my_metadata, ...})
+    """
+
+    @classmethod
+    @abstractmethod
+    def namespace(cls) -> str:
+        raise NotImplementedError()
+
+    def keys(self) -> AbstractSet[str]:
+        dictionary = self.dict()
+        return {
+            f"{self.namespace()}/{key}" for key in dictionary.keys() if dictionary[key] is not None
+        }
```

Comment:
oh dope we can use `computed_field` in pydantic to memoize https://github.com/pydantic/pydantic/pull/5502

(not critical here, but very happy that pattern exists)